### PR TITLE
Remove dead and unused code

### PR DESCRIPTION
**Pull request by fppt**
# Why is this PR needed?

It is needed as routine maintenance so we can keep our code base clean and easier to work with.

# What does the PR do?

Removes Methods which are unused.
Annotates some methods which are part of the public API.
Annotates some methods which are used indirectly, for example the OLAP constructors

# Does it break backwards compatibility?

Nope. Everything being removed should not be in use.

# List of future improvements not on this PR

Automatically pickup dead code and break the build because of it but I have not figured out how to achieve this.

# Note:

Because I had to go through a lot of unfamiliar code I ended up tagging comments for different people to address. I would appreciate it if you guys could help me address these unknowns.

### CODE REVIEW

File: grakn-core/src/main/java/ai/grakn/graql/GetQuery.java [hunk 1]
```diff
@@ -48,8 +47,9 @@
     /**
      * Get the transaction this query is running against, if it is set
      */
-    @CheckReturnValue
-    Optional<GraknTx> tx();
+    @SuppressWarnings("unsued") //TODO: @Felix, What's the deal here? Seems totally unused
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/GetQuery.java]: @aelred

File: grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java [hunk 2]
```diff
@@ -153,6 +153,9 @@
     @CheckReturnValue
     Atomic inferTypes();
 
+    //TODO @Kapser: This is only used through it's implementation class. Making the method here unneeded
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java]: @kasper-piskorski

File: grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java [hunk 3]
```diff
@@ -37,8 +37,11 @@
      * @param tx the {@link GraknTx} to execute the compute query on
      * @return a compute query builder with the graph set
      */
-    @CheckReturnValue
-    CentralityQueryBuilder withTx(GraknTx tx);
+    //TODO: @Jason. What's the deal with this method?
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java]: @duckofyork

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/spanningtree/ExclusiveEdge.java [hunk 4]
```diff
@@ -44,6 +44,7 @@ private ExclusiveEdge(DirectedEdge<V> edge, List<DirectedEdge<V>> excluded, doub
         return new ExclusiveEdge<>(edge, excluded, weight);
     }
 
+    @SuppressWarnings("unused")//TODO: @Jason, how is this used? I can't find a reference to it?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/spanningtree/ExclusiveEdge.java]: @duckofyork

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractCentralityQuery.java [hunk 5]
```diff
@@ -62,6 +62,8 @@ void initSubGraph() { //TODO: REMOVE THIS METHOD
         includeAttribute = true;
     }
 
+    //TODO: @Jason. This is only ever used by a test. Is it part of the public api? Why do we need this method?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractCentralityQuery.java]: @duckofyork

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java [hunk 6]
```diff
@@ -45,12 +45,16 @@
     Set<Label> statisticsResourceLabels = new HashSet<>();
     Set<Type> statisticsResourceTypes = new HashSet<>();
 
+    //TODO: @Jason. This is only ever used by a test. Is it part of the public api? Why do we need this method?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java]: @duckofyork

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java [hunk 7]
```diff
@@ -45,12 +45,16 @@
...
+    @SuppressWarnings("unused")
     public V of(String... statisticsResourceTypeLabels) {
         this.statisticsResourceLabels =
                 Arrays.stream(statisticsResourceTypeLabels).map(Label::of).collect(Collectors.toSet());
         return (V) this;
     }
 
+    //TODO: @Jason. This is only ever used by a test. Is it part of the public api? Why do we need this method?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java]: @duckofyork

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java [hunk 8]
```diff
@@ -145,6 +145,9 @@ public void add(Cache<Q, T> c2){
      * @param answers answer stream of the query
      * @return lazy iterator of updated answers
      */
+    //TODO: @Kasper Is this abstract method on `Cache` really needed?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: @kasper-piskorski

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java [hunk 9]
```diff
@@ -153,9 +156,18 @@ public void add(Cache<Q, T> c2){
      * @return unified cached answers
      */
     public abstract T getAnswers(Q query);
+    //TODO: @Kasper Is this abstract method on `Cache` really needed?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: @kasper-piskorski

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java [hunk 10]
```diff
@@ -153,9 +156,18 @@ public void add(Cache<Q, T> c2){
...
+    //      It's only ever called through LazyQueryCache
+    @SuppressWarnings("unused")
     public abstract Pair<T, MultiUnifier> getAnswersWithUnifier(Q query);
     public abstract Stream<Answer> getAnswerStream(Q query);
+    //TODO: @Kasper Is this abstract method on `Cache` really needed?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: @kasper-piskorski

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java [hunk 11]
```diff
@@ -153,9 +156,18 @@ public void add(Cache<Q, T> c2){
...
+    //      It's only ever called through LazyQueryCache
+    @SuppressWarnings("unused")
     public abstract Pair<Stream<Answer>, MultiUnifier> getAnswerStreamWithUnifier(Q query);
+    //TODO: @Kasper Is this abstract method on `Cache` really needed?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: @kasper-piskorski

File: grakn-core/src/main/java/ai/grakn/graql/GetQuery.java (hunk 1 above)
Comment (aelred) [reply grakn-core/src/main/java/ai/grakn/graql/GetQuery.java]: Just a getter. Arguably part of the API, but I'm not sure either way. Maybe tag as deprecated?

File: grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java [hunk 12]
```diff
@@ -37,8 +37,11 @@
...
+    // I remove it and it's implementation and everything still works
+    //@SuppressWarnings("unused")
+    //@CheckReturnValue
+    //CentralityQueryBuilder withTx(GraknTx tx);
```
Comment (duckofyork) [inline grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java]: The idea was, the user can build a graql query without a tx at the beginning, as long as a tx is provided before execution. So this method is everywhere .

File: grakn-core/src/main/java/ai/grakn/graql/GetQuery.java (hunk 1 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/graql/GetQuery.java]: Can't we just kill it? I doubt it's being used externally?

File: grakn-bootup/src/main/java/ai/grakn/bootup/ProcessHandler.java [hunk 13]
```diff
@@ -19,20 +19,29 @@
 package ai.grakn.bootup;
 
 /**
+ * The methods of this interface are used to define a common expectation of what processes should offer.
+ * The actual {@link ProcessHandler} interface is never directly accessed.
+ * Therefore the need for the suppressions below
```
Comment (aelred) [inline grakn-bootup/src/main/java/ai/grakn/bootup/ProcessHandler.java]: I'd rather remove unused interfaces like this - but I think this is a personal opinion so keeping it is all good.

File: grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java [hunk 14]
```diff
@@ -178,7 +180,7 @@ public static Builder newBuilder() {
         return new Builder();
     }
 
-    //TODO: Remove this method used only by docs tests
+    @API
```
Comment (aelred) [inline grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java]: I'd deprecate this (or keep the TODO), since we wanted to remove it but it's used by the doc tests.

File: grakn-core/src/main/java/ai/grakn/GraknTx.java [hunk 15]
```diff
@@ -320,6 +320,7 @@
      * @return true if the current transaction is read only
      */
     @CheckReturnValue
+    @API
     boolean isReadOnly();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/GraknTx.java]: Since we have an opportunity to define our API now, how about a method that returns the `GraknTxType` instead of this more-specific one?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/spanningtree/ExclusiveEdge.java [hunk 16]
```diff
@@ -44,6 +44,7 @@ private ExclusiveEdge(DirectedEdge<V> edge, List<DirectedEdge<V>> excluded, doub
...
     public static <T> ExclusiveEdge<T> of(DirectedEdge<T> edge, double weight) {
         return ExclusiveEdge.of(edge, ImmutableList.of(), weight);
     }
```
Comment (duckofyork) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/spanningtree/ExclusiveEdge.java]: You are right! It's no longer used.

File: grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java [hunk 17]
```diff
@@ -206,6 +207,7 @@
     /**
      * Closes the root session this graph stems from. This will automatically rollback any pending transactions.
      */
+    @API
     void closeSession();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java]: No one is using this and it's on an admin interface! Sounds like killing time.

File: grakn-engine/src/main/java/ai/grakn/engine/controller/response/Keyspace.java [hunk 18]
```diff
@@ -37,45 +38,53 @@
 @AutoValue
 public abstract class Keyspace implements Jacksonisable {
 
+    @API
```
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/controller/response/Keyspace.java]: Kind of neat how we are accidentally marking what parts of our REST API are API.

File: grakn-graql-shell/src/main/java/ai/grakn/graql/JsonSession.java [hunk 19]
```diff
@@ -145,6 +145,7 @@ void sendJson(Json json) throws WebSocketException, IOException {
     }
 
     @OnWebSocketClose
+    @SuppressWarnings("unused") //Used through socket connection.
```
Comment (aelred) [inline grakn-graql-shell/src/main/java/ai/grakn/graql/JsonSession.java]: IntelliJ magic: you can actually tell IntelliJ that `@OnWebsocketClose` marks an entry point if you want instead.

File: grakn-graql-shell/src/main/java/ai/grakn/graql/internal/shell/animalia/chordata/mammalia/artiodactyla/hippopotamidae/Hippopotamus.java [hunk 20]
```diff
@@ -35,5 +35,6 @@
 
     void submerge();
 
+    @SuppressWarnings("unused")
     boolean isHungryHungry();
```
Comment (aelred) [inline grakn-graql-shell/src/main/java/ai/grakn/graql/internal/shell/animalia/chordata/mammalia/artiodactyla/hippopotamidae/Hippopotamus.java]: `@API` for sure

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java [hunk 21]
```diff
@@ -539,32 +539,27 @@ else if(compatibleTypesFromRoles.isEmpty() || types.isEmpty()){
      * {@link EntityType}s only play the explicitly defined {@link Role}s (not the relevant part of the hierarchy of the specified {@link Role}) and the {@link Role} inherited from parent
      * @return list of {@link RelationshipType}s this atom can have ordered by the number of compatible {@link Role}s
      */
-    public ImmutableList<Type> inferPossibleTypes(Answer sub) {
+    public ImmutableList<Type> inferPossibleTypes() {
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: I tried to remove this and stuff started breaking, so @kasper-piskorski suggested I roll it back. But apparently it's working for you soooo

File: grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java [hunk 22]
```diff
@@ -699,7 +699,6 @@ protected void clearGraph() {
         getTinkerPopGraph().traversal().V().drop().iterate();
     }
 
-    @Override
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java]: This method is still on the interface, so this should stay.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractCentralityQuery.java [hunk 23]
```diff
@@ -62,6 +62,8 @@ void initSubGraph() { //TODO: REMOVE THIS METHOD
...
+    @SuppressWarnings("unused")
     public V of(String... ofTypeLabels) {
         return of(Arrays.stream(ofTypeLabels).map(Label::of).collect(Collectors.toSet()));
     }
```
Comment (duckofyork) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractCentralityQuery.java]: It's part of the public API. The idea was to have two methods for specifying types (subgraph), one takes string, one takes label. I can remove one of the two if you want.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java [hunk 24]
```diff
@@ -45,12 +45,16 @@
...
+    @SuppressWarnings("unused")
     public V of(Collection<Label> statisticsResourceLabels) {
         this.statisticsResourceLabels = Sets.newHashSet(statisticsResourceLabels);
         return (V) this;
```
Comment (duckofyork) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/AbstractStatisticsQuery.java]: Ditto

File: grakn-core/src/main/java/ai/grakn/graql/GetQuery.java (hunk 1 above)
Comment (aelred) [reply grakn-core/src/main/java/ai/grakn/graql/GetQuery.java]: I'm good with that! We toe the line of "breaking change", but you're right it's not likely to be used. Plus, we haven't defined our API yet :P

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java (hunk 11 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: @aelred had the same questions in his cleanup PR :D

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java (hunk 21 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: Well in general things shouldn't break as this deals with type inference. It's just that less will be inferred. I will double check. Probably on the finance dataset.

Comment (fppt) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: `Answer sub` was never used anywhere in the chain so that's why I removed it and then when you do that the whole method looks alot like the existing one.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java [hunk 25]
```diff
@@ -360,17 +360,6 @@ public GetQuery getQuery() {
...
-    public ImmutableMap<Var, Type> getVarTypeMap(Answer sub) {
-        return ImmutableMap.copyOf(getVarTypeMap(
-                Stream.concat(
-                        getAtoms(IsaAtom.class),
-                        inferEntityTypes()
-                )
-                )
-        );
-    }
-
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java]: ok it probably should be `inferEntityTypes(sub)` here so that the sub is actually used in all the infer* methods

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java (hunk 11 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: these are all implemented in `QueryCache` and `LazyQueryCache` so wanted to have a common parent

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java [hunk 26]
```diff
@@ -539,32 +539,27 @@ else if(compatibleTypesFromRoles.isEmpty() || types.isEmpty()){
...
                     .filter(at -> !Sets.intersection(at.getVarNames(), untypedRoleplayers).isEmpty())
                     .collect(toSet());
 
             ImmutableList.Builder<Type> builder = ImmutableList.builder();
-            //prioritise relations with higher chance of yielding answers
             compatibleConfigurations.asMap().entrySet().stream()
-                    //prioritise relations with more allowed roles
-                    .sorted(Comparator.comparing(e -> -e.getValue().size()))
-                    //prioritise relations with number of roles equal to arity
-                    .sorted(Comparator.comparing(e -> e.getKey().relates().count() != getRelationPlayers().size()))
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: why is this removed? i know that intelliJ doesn't like it but I don't get why

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/plan/SimplePlanner.java [hunk 27]
```diff
@@ -114,16 +110,4 @@
...
-     * @param query for which the plan should be constructed
-     * @return list of atoms in order they should be resolved using {@link GraqlTraversal}.
-     */
-    static public ImmutableList<Atom> plan(ReasonerQueryImpl query){
-        return ImmutableList.<Atom>builder().addAll(
-                query.selectAtoms().stream()
-                        .sorted(Comparator.comparing(at -> -at.baseResolutionPriority()))
-                        .iterator())
-                .build();
-    }
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/plan/SimplePlanner.java]: this our old planner, it is not used in the code however it is useful for debugging, it uses the priority function from `Atom`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/plan/GraqlTraversalPlanner.java [hunk 28]
```diff
@@ -55,17 +55,6 @@
...
-     * @return list of atoms in order they should be resolved using {@link GraqlTraversal}.
-     */
-    public static ImmutableList<Atom> plan(ReasonerQueryImpl query){
-        List<Atom> atoms = query.getAtoms(Atom.class)
-                .filter(Atomic::isSelectable)
-                .collect(Collectors.toList());
-        return planFromTraversal(atoms, query.getPattern(), query.tx());
-    }
-
     /**
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/plan/GraqlTraversalPlanner.java]: also simpler plan version, not used but potentially useful when debugging

File: grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java (hunk 2 above)
Comment (kasper-piskorski) [reply grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java]: can remove

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java (hunk 26 above)
Comment (fppt) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: IntelliJ said it was redundant and did not affect processing. Apparently java is actually smart enough to optimise these cases by itself. So these lines were (as far as I understand it) useless.

Comment (aelred) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: IntelliJ probably thinks the sorts aren't stable, so `stream.sorted(byFoo).sorted(byBar)` is the same as `stream.sorted(byBar)`.

@kasper-piskorski if you want to sort by Foo first, then by Bar second try this:

```
stream.sorted(byFoo.thenComparing(byBar))
```

I don't know if it will make a difference, but it won't make things worse and at least IntelliJ will stop complaining 😛

File: grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java (hunk 12 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/graql/analytics/CentralityQueryBuilder.java]: @duckofyork is the idea still valid though? I am not sure if we sitll aiming to provide such on the API?

File: grakn-bootup/src/main/java/ai/grakn/bootup/ProcessHandler.java (hunk 13 above)
Comment (fppt) [reply grakn-bootup/src/main/java/ai/grakn/bootup/ProcessHandler.java]: Killing it.

File: grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java (hunk 14 above)
Comment (fppt) [reply grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java]: This is actually the way we want users to interface with the BEC programatically. So it's actual public API . . . according to the docs at least.

File: grakn-core/src/main/java/ai/grakn/GraknTx.java (hunk 15 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/GraknTx.java]: Good catch! Will do sir!

File: grakn-graql-shell/src/main/java/ai/grakn/graql/internal/shell/animalia/chordata/mammalia/artiodactyla/hippopotamidae/Hippopotamus.java (hunk 20 above)
Comment (fppt) [reply grakn-graql-shell/src/main/java/ai/grakn/graql/internal/shell/animalia/chordata/mammalia/artiodactyla/hippopotamidae/Hippopotamus.java]: Wow you right. Horrible mistake on my side 😄

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java (hunk 11 above)
Comment (fppt) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/cache/Cache.java]: If they not accessed through the abstraction then it's an unneeded inheritance. So I am going to kill it. Just shout if you totally disagree.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java (hunk 25 above)
Comment (fppt) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java]: But it wasn't in use though? I can add it back but nothing used it.

Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java]: `getVarTypeMap(sub`) is referenced in the infer functions in `RelationshipAtom` and others

**Comment by duckofyork** [comment]
It is still valid, as we do need the method for building a query without a tx. Though we could have something simpler: `query.execute(tx)` or `tx.execute(query)`.

I think we should keep it for now. Felix and I can clean this up in the future.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java (hunk 25 above)
Comment (fppt) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java]: Yes I saw that, but `sub` was never used so I removed the parameter. At that stage I noticed that `public ImmutableMap<Var, Type> getVarTypeMap()` and `public ImmutableMap<Var, Type> getVarTypeMap(Answer sub) ` pretty much did the same thing.

So now everything calls `getVarTypeMap()` and it all still appears to be working. 

How are these methods different from each other that it's worth keeping both?

**Comment by fppt** [comment]
@aelred @duckofyork @kasper-piskorski Could I get a re-review guys?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java (hunk 25 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java]: the one with sub infers types from instance ids

File: grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java [hunk 29]
```diff
@@ -120,19 +121,19 @@ private BatchExecutorClient(Builder builder) {
     /**
      * Will block until there is space for the query to be submitted
      */
-    public Observable<QueryResponse> add(Query<?> query, Keyspace keyspace) {
+    public Observable add(Query<?> query, Keyspace keyspace) {
```
Comment (aelred) [inline grakn-client/src/main/java/ai/grakn/client/BatchExecutorClient.java]: You missed a spot with your raw type fixing! This one is important because it's the entry-point to using the `BatchExecutorClient`.

If we don't want to expose the `QueryResponse` type (since it is useless), we can make this return `Observable<?>`.

File: grakn-client/src/main/java/ai/grakn/client/QueryResponse.java [hunk 30]
```diff
@@ -15,42 +15,19 @@
...
-    }
-
-    public Json getJsonResponse() {
-        return jsonResponse;
-    }
-
-    public Query<?> getQuery() {
-        return query;
-    }
+    public static final QueryResponse INSTANCE = new QueryResponse();
```
Comment (aelred) [inline grakn-client/src/main/java/ai/grakn/client/QueryResponse.java]: Yay! It's the unit type :D

File: grakn-test/test-integration/src/test/java/ai/grakn/test/graql/shell/GraqlShellIT.java [hunk 31]
```diff
@@ -644,7 +644,7 @@ private void assertShellMatches(List<String> arguments, Object... matchers) thro
 
         ImmutableSet<String> noPromptArgs = ImmutableSet.of("-e", "-f", "-b", "-v", "-h");
         if (Sets.intersection(Sets.newHashSet(arguments), noPromptArgs).isEmpty()) {
-            // Remove first four lines containing license and last line containing prompt
+            // Remove first five lines containing license and last line containing prompt
             outputLines = outputLines.subList(4, outputLines.size() - 1);
```
Comment (aelred) [inline grakn-test/test-integration/src/test/java/ai/grakn/test/graql/shell/GraqlShellIT.java]: But it says 4! Who is the liar??

Comment (fppt) [reply grakn-test/test-integration/src/test/java/ai/grakn/test/graql/shell/GraqlShellIT.java]: Debugging some stupid stuff will fix

File: grakn-core/src/main/java/ai/grakn/graql/analytics/CorenessQuery.java [hunk 32]
```diff
@@ -50,13 +50,6 @@
...
     CorenessQuery of(String... ofTypeLabels);
 
-    /**
-     * @param ofLabels a collection of types in the subgraph to compute coreness of. By default the coreness of all
-     *                 entities and attributes are computed.
-     * @return a CorenessQuery with the subTypeLabels set
-     */
-    CorenessQuery of(Collection<Label> ofLabels);
-
     /**
```
Comment (duckofyork) [inline grakn-core/src/main/java/ai/grakn/graql/analytics/CorenessQuery.java]: I think we should keep this method

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java (hunk 26 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: From my understanding the sorting here is stable. I have just tested with markit. Removing the extra sorting makes some of the query execution times go through the roof xD

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java [hunk 33]
```diff
@@ -574,7 +569,6 @@ else if(compatibleTypesFromRoles.isEmpty() || types.isEmpty()){
                                 rs.stream().flatMap(Role::playedByTypes).filter(typesFromNeighbour::contains).count()
                         );
                     })
-                    .sorted(Comparator.comparing(p -> -p.getValue()))
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: this needs to stay as well
